jsv2/jetstream/errors.go

```diff
@@ -0,0 +1,189 @@
+
+	// ErrMsgAlreadyAckd is returned when attempting to acknowledge message more than once.
+	ErrMsgAlreadyAckd JetStreamError = &jsError{message: "message was already acknowledged"}
+
+	// ErrNoStreamResponse is returned when there is no response from stream (e.g. no responders error).
+	ErrNoStreamResponse JetStreamError = &jsError{message: "no response from stream"}
+
+	// ErrNotJSMessage is returned when attempting to get metadata from non JetStream message .
+	ErrNotJSMessage JetStreamError = &jsError{message: "not a jetstream message"}
+
+	// ErrInvalidStreamName is returned when the provided stream name is invalid (contains '.').
+	ErrInvalidStreamName JetStreamError = &jsError{message: "invalid stream name"}
+
+	// ErrInvalidConsumerName is returned when the provided consumer name is invalid (contains '.').
+	ErrInvalidConsumerName JetStreamError = &jsError{message: "invalid consumer name"}
+
+	// ErrNoMessages is returned when no messages are currently available for a consumer.
+	ErrNoMessages = &jsError{message: "no messages"}
+
+	// ErrHandlerRequired is returned when no handler func is provided in Stream()
+	ErrHandlerRequired = &jsError{message: "handler cannot be empty"}
+
+	// ErrEndOfData is returned when iterating over paged API from JetStream reaches end of data
+	ErrEndOfData = errors.New("nats: end of data reached")
+
+	// ErrNoHeartbeat is received when no message is received in IdleHeartbeat time (if set)
+	ErrNoHeartbeat = &jsError{message: "no heartbeat received, canceling subscription"}
+
+	// ErrConsumerHasActiveSubscription is returned when a consumer is already subscribed to a stream
+	ErrConsumerHasActiveSubscription = &jsError{message: "consumer has active subscription"}
+
+	// ErrMsgNotBound is returned when given message is not bound to any subscription
+	ErrMsgNotBound = &jsError{message: "message is not bound to subscription/connection"}
+
+	// ErrMsgNoReply is returned when attempting to reply to a message without a reply subject
+	ErrMsgNoReply = &jsError{message: "message does not have a reply"}
+
+	// ErrMsgDeleteUnsuccessful is returned when an attempt to delete a message is unsuccessful
+	ErrMsgDeleteUnsuccessful = &jsError{message: "message deletion unsuccessful"}
+
+	// ErrAsyncPublishReplySubjectSet is returned when reply subject is set on async message publish
+	ErrAsyncPublishReplySubjectSet = &jsError{message: "reply subject should be empty"}
+
+	// ErrTooManyStalledMsgs is returned when too many outstanding async messages are waiting for ack
+	ErrTooManyStalledMsgs = &jsError{message: "stalled with too many outstanding async published messages"}
+
+	// ErrInvalidOption is returned when there is a collision between options
+	ErrInvalidOption = &jsError{message: "invalid jetstream option"}
+)
+
+// Error prints the JetStream API error code and description
+func (e *APIError) Error() string {
+	return fmt.Sprintf("nats: API error %d: %s", e.ErrorCode, e.Description)
```

Comment:
Thinking now that maybe we should add the `Code` as well here besides the error code since that can give more of the semantic meaning right away, like a bad KV update becomes would be a bad request with `code=400` and  `err_code=10071 description='wrong last sequence: 1'`